src/librustc/ich/impls_ty.rs

```diff
@@ -384,6 +384,30 @@ for ::middle::const_val::ConstVal<'gcx> {
     }
 }
 
+impl<'a, 'gcx> HashStable<StableHashingContext<'a>>
+for ::middle::const_val::ConstValue<'gcx> {
```

Comment:
Please do not add more things to that module. It should be removed already. Everything constant belongs in miri.

Reply:
It almost seems like you should be able to make miri's `Value` generic over the allocation information in the `ByRef` case.

Reply:
I agree. Also it would probably suffice to return the `Allocation` directly instead of an interned one.

Reply:
> Please do not add more things to that module. It should be removed already. Everything constant belongs in miri.

Do you want to move `ty::Const`, `ConstVal` and `ConstValue` to a new module `librustc\mir\interpret\const.rs`?

> It almost seems like you should be able to make miri's Value generic over the allocation information in the ByRef case.

I don't think there's anything to be gained by doing that. I'd rather just get rid of miri's Value type. Representing values by pointers to values in some cases and just values in other cases is confusing.

> Also it would probably suffice to return the Allocation directly instead of an interned one.

I don't know what you mean here, but `ConstValue` can't store an `Allocation` directly, since it constants are in the dropless interner. Interning here also makes type folding cheaper.

Reply:
@Zoxc `ConstValue` should be right next to the existing miri `Value`, while `ConstVal` should be `ConstKind` right next to the definition of `ty::Const`.